Versioning (#17)
neighborhoods/ThrowableDiagnosticComponent: README.md

# Throwable Diagnostic Component

A Service Agnostic Component for diagnosing `Throwables`. 

The `ThrowableDiagnostic` can determine if a `Throwable` happened because of a permanent or transient fault (network issues, database down, 3rd party service unavailable).

Every `Throwable` is by default considered permanent, but certain `Throwables` are known to be transient. With custom diagnostic logic such cases can be identified.

Diagnostic logic for common cases like RDBMS and AWS are available out of the box. Defining your own diagnostic logic is straightforward.

## Install

Via Composer

```bash
$ composer require neighborhoods/throwable-diagnostic-component
```

> ###Versioning

Reviewer: ```suggestion
### Versioning
```

Don't confuse releases on [packagist](https://packagist.org/packages/neighborhoods/throwable-diagnostic-component), e.g. 3.0.0, with versions of components contained within this package, e.g. AwsSqsV1. One release may contain multiple versions of the same component.  
Parts of your code may use older versions of components, while others use the latest.

### Upgrade

For upgrading from release 3 to release 4 follow the [Upgrade Guide](docs/UpgradeGuide.md).

## Usage

Components fall into one of two categories
* Throwable Diagnostic
* Throwable Diagnostic Decorator

> The Throwable Diagnostic entails the concept of how this works. The Throwable Diagnostic mostly contains interfaces. It determines how user code diagnoses a Throwable, how the decorators look like, how to define the stack of decorators to be used. You may think of it as a framework/skeleton/scaffolding.  

Reviewer: ```suggestion
The Throwable Diagnostic defines the core concept of how this works. The Throwable Diagnostic mostly contains interfaces. It determines how user code diagnoses a Throwable, how the decorators look, and how to define the stack of decorators to be used. You may think of it as a framework/skeleton/scaffolding.  
```

There might be multiple versions of throwable diagnostic, e.g. ThrowableDiagnosticV1 and ThrowableDiagnosticV2.

> A Throwable Diagnostic Decorator is in charge to diagnose certain `Throwables`. For example the `GuzzleV1` component has a decorator which decides if an exception originating from guzzle is transient or not. During the diagnostic process a `Throwable` can be inspected by multiple decorators, each of which may decide that the `Throwable` is transient, permanent or should be passed to the next decorator.

Reviewer: ```suggestion
A Throwable Diagnostic Decorator is in charge of diagnosing certain `Throwables`. For example, the `GuzzleV1` component has a decorator which decides if an exception originating from guzzle is transient or not. During the diagnostic process a `Throwable` can be inspected by multiple decorators, each of which may decide that the `Throwable` is transient, permanent, or should be passed to the next decorator.
```

(that last comma is an "oxford comma". i prefer these, but they are definitely not required as a rule and i try not to suggest things that aren't necessary. however, here i feel like it helps clarify the meaning of the sentence.

A throwable diagnostic decorator is made for a specific version of Throwable Diagnostic. All decorators compatible with ThrowableDiagnosticV1 are placed in the `src/ThrowableDiagnosticV1Decorators` folder.  

> There are decorators for different `Throwables`. Diagnostic logic for certain `Throwables` may have multiple versions, e.g. GuzzleV1, GuzzleV2. Use the latest version if possible.  

Reviewer: ```suggestion
There are decorators for different `Throwables`. Diagnostic logic for certain `Throwables` may have multiple versions, e.g. GuzzleV1, GuzzleV2. Use the latest version if possible. The versions here do not correspond to versions of guzzle, rather versions of the specific decorator.
```

You can implement your own decorator to properly handle `Throwables` which are specific to your own code, or a package for which a decorator is not available.

> For usage details and list of predefined decorators please refer to the corresponding Throwable Diagnostic version.

Reviewer: :+1:
